### ISSUE TITLE
[frio] WYSIWYG BBCode/Markdown editor

### ISSUE THREAD

**Issue by Extarys** [friendica/friendica#9848]
### Is the feature request related to a problem? Please describe.

BBCode is not user-friendly. It's awesome though.

It requires prior knowledge, there is a learning curve (small, but still). Although it is very flexible as we can add any tag we want, users need to learn them.

### Describe the feature you'd like

I dig into old issues and found #2810

I'd like to make an editor that would abstract any type of code. A more advanced user could activate the full functionality of the editor and configure it how they like, but without the option, a plain and simple text box will be shown, maybe with a small button to temporary enable the editor.

When the editor is activated, a toolbar would be displayed above the textbox with the option to swich to code (like how it is now).

Markdown could be implemented later, but it wouldn't be the starting point. I thought of this, well, as soon as I installed friendica but also because I'd like to have an emoji picker (#9298). That would require to add a new button. Hence my thought about a WYSIWYG editor.

In the background, everything will be converted to BBCode.

### Additional context

This would provide a lower level of entry for new users with a more intuitive posting mechanism. Also, I'm exploring file uploads and I would have to rewrite most part of it once this is implemented. This seems the first logical step in my adventure.

Beware: This is a discussion/approval request and this won't be merged tomorrow. A friend here said to me I should be careful not to eat more than I can chew :smiley: This will act as a reminder and task tracker since it is a big chunk and will clutter my UI improvement issue [#9787].

**Comment by MrPetovan** [comment]
I have strong mostly negative feelings about WYSIWYG editors. Either they are too simple and can't cover all the underlying markup language features, or they are too complex and they require a separate learning curve and take a lot of screen real estate, and that's even before verifying the touchscreen/mobile support. And, more often than not, WYSINWYG (What You See Is Not What You Get).

I personally end up switching to the source code mode (when it exists!) whenever I'm faced with a WYSIWYG editor.

Now I know that I'm a developer so I'm used to source code, whether it is BBCodes, Markdown, PHP, Javascript, CSS or HTML, but I remember phpBB forums populated by non-IT people that were successfully using BBCodes because that's what was available. So while I'm in a privileged situation, I do not believe it makes WYSIWYG editors necessary or even desirable for non-IT people. However it will introduce on its own a whole host of additional support needs that overcome the estimated benefits for the project.

My bottom line is, on the subject of the post/comment box improvements, there are other pending feature requests that are far more impactful in my eyes than a WYSIWYG editor, beginning with the [drag-and-drop feature to share pictures](https://github.com/friendica/friendica/issues/6249) (#6249), or the custom emoji picker (every OS has a native emoji picker of their own, so it doesn't seem relevant to me to redo another one).

**Comment by Extarys** [comment]
I do agree, but we're not in the 90s anymore and "regular" people are having higher standards now.

I think I didn't express this issue properly, but I don't want to focus on WYSIWYG right away, but start with a redesign of the toolbar/textarea section. Adding the emoji picker next to all other current buttons seems like the path of least resistance while cluttering the UI. Eventually it will need to be done.

![image](https://user-images.githubusercontent.com/11581433/105499126-02ae7b00-5c8f-11eb-85cc-36fe2a63b2a0.png)

I agree about WYSI**N**WYG part :joy: and I'm also working on the other stuff. I'm afraid I'll have to redo some stuff later if we decide to redesign the textarea section.

**Comment by tiotrom** [comment]
Same way that wordpress does I think it fits both worlds. A tabbed content creator. One tab is for "code" and one for wysiwyg, with the latter being not very pretentious, just simple text editing: bold, italic, position, quote, etc.. I think it would be very helpful to have a wysiwyg composer. I understand @MrPetovan but the vast majority of people need a lot less friction and  wysiwyg could be focused on very basic editing. For those who want a lot more control, they have to learn to use the "code" tab. ;)

**Comment by Extarys** [comment]
That is my general idea too. Thanks for chiming in!

Anyway, any changes will be subject to a PR and a discussion about it.

**Comment by utzer** [comment]
Please, can we have a more improved editor, there is need for it.

**Comment by MrPetovan** [comment]
"improved editor" doesn't mean much, what is your/the main need that should be addressed? I will tackle the file upload for next release but otherwise?

**Comment by utzer** [comment]
Just as @Extarys describes it in the top. ;-) 

Or well, just some of the formattings available as buttons instead of the BBCode. I never know what formatting and tags are there, there is no list showing up for BBCode, maybe we can have some way to search for BBCode tags, like typing the sqaure bracket and then any work it will suggest BBCode tags that fit? Most a logical, but not in other languages. 

I mean type the bracket and then you can search for some keywords or so? No idea how it could work, but also I like the idea to have buttons of the most common tags/formatting, many people would like that I think.

**Comment by MrPetovan** [comment]
If you start typing an opening bracket character, it already shows a list of available BBCodes. Typing more letters refines the list based on the prefix.

**Comment by utzer** [comment]
> Boy have I good news for you, if you start typing an opening bracket character, it already shows a list of available BBCodes. Typing more letters refines the list based on the prefix.

but not a complete list, there is more tags. I wonder when this was added, I used to type open bracket a lot, but never got the list, till now.

**Comment by tobiasd** [comment]
TinyMCE (the WYSIWYG editor Friendica once used but everyone did not like) was removed 2017. Before that the completion drop-down for BBCode was added and well tested IIRC.

**Comment by MrPetovan** [comment]
The complete list cannot be shown in an auto-completion popup, the current system based on the first few letters you type after the bracket is the best we can do in this context. The complete list of tags is available in the node documentation, although we don't link to it from any post/comment form and we probably should.

**Comment by SpencerDub** [comment]
> (every OS has a native emoji picker of their own, so it doesn't seem relevant to me to redo another one)

Just to speak to this: The appeal of an inline emoji picker like [Emoji Mart](https://github.com/missive/emoji-mart) for me is that it allows easy browsing of the server's custom emoji. The ability to easily manage custom emoji as an admin, and use them as a user, is a feature I wish Friendica had—the current options (the custom smilies add-ons or the existing smiley picker in the composer) are difficult to use compared to my experience with Mastodon.

**Comment by MrPetovan** [comment]
This parenthesis wasn't meant to dismiss emoji pickers in general, I wanted to put the emphasis on a potential _custom_ emoji picker which is dearly needed to make any use of the Friendica custom emojis.

**Comment by utzer** [comment]
> The complete list cannot be shown in an auto-completion popup, the current system based on the first few letters you type after the bracket is the best we can do in this context. The complete list of tags is available in the node documentation, although we don't link to it from any post/comment form and we probably should.

Yes, so you see the short coming as well, but I'll give up, I can understand that this is not a thing anyone want so add. 

@Extarys I hope you make something that is better!
